### PR TITLE
Execsan integration 1

### PR DESCRIPTION
**Pull request by Alan32Liu**
Part 1 of integrating `execSan` into `clusterfuzz`:
1. Refactor `UnshareProcessRunner` into `ModifierProcessRunner`
2. Prepend fuzzing commands with the sanitiser.

I was using the name `injSan` instead of `execSan` as we may need to add other injection detection later.

TODOs:
1. Build sanitiser in `OSS-FUZZ`
2. Capture error messages from the sanitiser
3. Roll out gradually.

### CODE REVIEW

File: src/clusterfuzz/_internal/system/new_process.py [hunk 1]
```diff
@@ -414,36 +421,37 @@ class UnicodeProcessRunner(UnicodeProcessRunnerMixin, ProcessRunner):
...
-
-    command.append(self._executable_path)
+    command = [self._executable_path]
     command.extend(self._default_args)
 
     if additional_args:
       command.extend(additional_args)
 
-    return command
+    return self.tool_prefix('unshare') + self.tool_prefix('inj') + command
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/system/new_process.py:449]: nit: rename inj to extra_sanitizers

File: src/clusterfuzz/_internal/system/new_process.py [hunk 2]
```diff
@@ -414,36 +421,37 @@ class UnicodeProcessRunner(UnicodeProcessRunnerMixin, ProcessRunner):
...
-  """ProcessRunner mixin which unshares."""
+class ModifierProcessRunnerMixin(object):
+  """ProcessRunner mixin with modifiers."""
+
+  def tool_prefix(self, tool):
+    """Prefix the command with a tool and its args"""
+    if not environment.get_value(f'USE_{tool.upper()}'):
+      return []
+
+    tool_path = environment.get_default_tool_path(tool)
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/system/new_process.py:432]: We'll also need to add a mechanism to download the binary if it is not installed in our image. e.g. for our "extra_sanitizers" we can hardcode it to a GCS bucket path to be downloaded as needed.

**Review by oliverchang — COMMENTED** [review]
nice!

**Review by oliverchang — APPROVED** [review]
LGTM, thanks! Could we just add some tests to `new_process_test` for this?

File: src/clusterfuzz/stacktraces/__init__.py [hunk 3]
```diff
@@ -823,6 +823,10 @@ def parse(self, stacktrace: str) -> CrashInfo:
           new_state='',
           reset=True)
 
+      # Shell bugs detected by extra sanitisers.
+      self.update_state_on_match(
+          EXTRA_SANITIZERS_SHELL_BUG_REGEX, line, state, new_type='Shell bug')
```
Comment (oliverchang) [inline src/clusterfuzz/stacktraces/__init__.py:828]: Can we rename "Shell bug" to "Command injection"?

File: src/clusterfuzz/_internal/system/new_process.py [hunk 4]
```diff
@@ -414,36 +430,47 @@ class UnicodeProcessRunner(UnicodeProcessRunnerMixin, ProcessRunner):
...
+    if not os.path.exists(tool_path) and tool in TOOL_URLS:
+      urllib.request.urlretrieve(TOOL_URLS.get(tool), tool_path)
+    if os.path.exists(tool_path) and not self.is_expected_mode(tool_path, tool):
+      os.chmod(tool_path, TOOL_MODES.get(tool))
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/system/new_process.py:450]: nit: no need for `TOOL_MODES`. we can just always use 0o755 for this.

File: src/clusterfuzz/_internal/tests/core/crash_analysis/stack_parsing/stack_analyzer_test.py [hunk 5]
```diff
@@ -3256,3 +3256,15 @@ def test_ignore_linux_gate(self):
     self._validate_get_crash_data(data, expected_type, expected_address,
                                   expected_state, expected_stacktrace,
                                   expected_security_flag)
+
+  def test_capture_shell_bug(self):
+    """Test capturing shell bugs detected by extra sanitizers"""
+    data = self._read_test_data('shell_bug.txt')
+    expected_type = 'Shell bug'
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/tests/core/crash_analysis/stack_parsing/stack_analyzer_test.py:3263]: Could we rename this to "Command injection" ?

**Review by oliverchang — APPROVED** [review]
Very nice!! LGTM with some minor comments.

Also, we need to add a condition here:

https://github.com/google/clusterfuzz/blob/8ae0c06ad52c392e8834d936156133ef6c8f3c89/src/clusterfuzz/_internal/system/environment.py#L151

If USE_EXTRA_SANITIZERS is set, we need to disable leak detection here.

**Comment by Alan32Liu** [comment]
I was trying to disable `LeakSan` in  one place (instead of repeating it at [417a5cc](https://github.com/google/clusterfuzz/pull/2702/commits/417a5cc1f05d8d817a938cb8ce20823808798818) and [7a275e4](https://github.com/google/clusterfuzz/pull/2702/commits/7a275e4157ac2c15e611fbb61603c36661040462)).
But I suppose there is no easy way to do so?

**Comment by oliverchang** [comment]
> I was trying to disable `LeakSan` in one place (instead of repeating it at [417a5cc](https://github.com/google/clusterfuzz/pull/2702/commits/417a5cc1f05d8d817a938cb8ce20823808798818) and [7a275e4](https://github.com/google/clusterfuzz/pull/2702/commits/7a275e4157ac2c15e611fbb61603c36661040462)). But I suppose there is no easy way to do so?

right we can just keep these two.

**Pull request merged**